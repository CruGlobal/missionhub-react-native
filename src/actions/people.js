import callApi, { REQUESTS } from './api';
import { PEOPLE_WITH_ORG_SECTIONS } from '../constants';
import { findAllNonPlaceHolders } from '../utils/common';

export function getMe() {
  return (dispatch) => {
    return dispatch(callApi(REQUESTS.GET_ME));
  };
}

export function getPerson(id) {
  return (dispatch) => {
    return dispatch(callApi(REQUESTS.GET_PERSON, { person_id: id }));
  };
}

export function getPeopleList() {
  return (dispatch) => {
    const query = {
      filters: {
        assigned_tos: 'me',
      },
    };
    return dispatch(callApi(REQUESTS.GET_PEOPLE_LIST, query));
  };
}

export function getMyPeople() {
  return (dispatch, getState) => {
    const peopleQuery = {
      filters: {
        assigned_tos: 'me',
      },
      include: 'reverse_contact_assignments,organizational_permissions',
    };

    return dispatch(callApi(REQUESTS.GET_PEOPLE_LIST, peopleQuery)).then((pResults) => {
      const myPeople = findAllNonPlaceHolders(pResults, 'person'); //todo do this by looking at something other than placeholder

      const personalPeople = myPeople.filter((person) => person.organizational_permissions.length === 0);
      const personalOrg = { people: personalPeople };

      if (getState().auth.isJean) {
        return dispatch(getMinistryPeople(myPeople, personalOrg));
      } else {
        return [ personalOrg ];
      }
    });
  };
}

function getMinistryPeople(myPeople, personalOrg) {
  const orgQuery = {
    filters: {
      assigned_tos: 'me',
    },
    include: '',
  };
  return (dispatch) => {
    return dispatch(callApi(REQUESTS.GET_MY_ORGANIZATIONS, orgQuery)).then((oResults) => {
      const ministryOrgs = oResults.findAll('organization');
      const ministryPeople = myPeople.filter((person) => person.organizational_permissions.length > 0);

      ministryOrgs.forEach((org) => org.people = []);

      ministryPeople.forEach((person) => {
        person.reverse_contact_assignments.forEach((ca) => {
          const org = ministryOrgs.find((o) => {
            const foundOrg = ca.organization;

            if (foundOrg) {
              const foundOrgPerm = person.organizational_permissions.find((op) => op.organization_id === foundOrg.id);

              return o.id === foundOrg.id && foundOrgPerm;
            }
            return false;
          });
          if (org) {
            org.people.push(person);
          }
        });
      });

      return dispatch({ type: PEOPLE_WITH_ORG_SECTIONS, myOrgs: [ personalOrg, ...ministryOrgs ] });
    });
  };
}

export function searchPeople(text, filters = {}) {
  return (dispatch) => {
    if (!text) {
      return Promise.reject('NoText');
    }

    let query = {
      q: text,
      filters: {},
    };
    if (filters.ministry) {
      query.organization_ids = filters.ministry.id;
    }
    if (filters.gender) {
      query.filters.gender = filters.gender.id;
    }
    if (filters.archived) {
      query.filters.archived = true;
    }
    if (filters.unassigned) {
      query.filters.unassigned = true;
    }
    if (filters.labels) {
      query.filters.label_ids = filters.labels.id;
    }
    if (filters.groups) {
      query.filters.group_ids = filters.groups.id;
    }
    if (filters.surveys) {
      query.filters.survey_ids = filters.surveys.id;
    }

    return dispatch(callApi(REQUESTS.SEARCH, query));
  };
}