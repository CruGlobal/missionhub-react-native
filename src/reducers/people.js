import { REHYDRATE } from 'redux-persist/constants';

import { REQUESTS } from '../actions/api';
import { DELETE_PERSON, LOGOUT, PEOPLE_WITH_ORG_SECTIONS, UPDATE_PERSON_ATTRIBUTES } from '../constants';
import { useFirstExists } from '../utils/common';

const initialState = {
  allByOrg: {},
};

function peopleReducer(state = initialState, action) {
  switch (action.type) {
    case REHYDRATE:
      const incoming = action.payload.people;
      if (incoming) {
        return {
          allByOrg: useFirstExists(incoming.allByOrg, state.allByOrg),
        };
      }
      return state;
    case REQUESTS.GET_PERSON.SUCCESS:
      const updatedPerson = action.results.response;
      return {
        ...state,
        allByOrg: updateAllPersonInstances(state.allByOrg, updatedPerson, true),
      };
    case UPDATE_PERSON_ATTRIBUTES:
      return {
        ...state,
        allByOrg: updateAllPersonInstances(state.allByOrg, action.updatedPersonAttributes),
      };
    case DELETE_PERSON:
      return {
        ...state,
        allByOrg: deletePersonInOrg(state.allByOrg, action.personId, action.personOrgId),
      };
    case LOGOUT:
      return initialState;
    case PEOPLE_WITH_ORG_SECTIONS:
      return {
        ...state,
        allByOrg: action.orgs,
      };
    default:
      return state;
  }
}

function updateAllPersonInstances(allByOrg, updatedPerson, replace = false) {
  return mapObject(allByOrg, ([ orgId, org ]) => ({
    [orgId]: {
      ...org,
      people: {
        ...org.people,
        ...org.people[updatedPerson.id] ?
          {
            ...replace ?
              { [updatedPerson.id]: updatedPerson } :
              { [updatedPerson.id]: { ...org.people[updatedPerson.id], ...updatedPerson } },
          } :
          {},
      },
    },
  }));
}

function deletePersonInOrg(allByOrg, deletePersonId, personOrgId) {
  return mapObject(allByOrg, ([ orgId, org ]) => ({
    [orgId]: {
      ...org,
      people: {
        ...orgId === personOrgId ?
          filterObject(org.people, ([ personId ]) => personId !== deletePersonId) :
          org.people,
      },
    },
  }));
}

function mapObject(obj, fn) {
  return Object.assign({}, ...Object.entries(obj).map(fn));
}

function filterObject(obj, fn) {
  return Object.assign({}, ...Object.entries(obj).filter(fn).map(([ id, element ]) => ({ [id]: element })));
}

export default peopleReducer;
