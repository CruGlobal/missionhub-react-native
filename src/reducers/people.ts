/* eslint complexity: 0 */

import { REQUESTS } from '../api/routes';
import {
  DELETE_PERSON,
  LOGOUT,
  PEOPLE_WITH_ORG_SECTIONS,
  LOAD_PERSON_DETAILS,
  UPDATE_PERSON_ATTRIBUTES,
  GET_ORGANIZATION_PEOPLE,
} from '../constants';

import { Organization } from './organizations';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Person = any;

export interface PeopleState {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  allByOrg: any;
}

const initialState: PeopleState = {
  allByOrg: {
    personal: { id: 'personal', people: {} },
  },
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default function peopleReducer(state = initialState, action: any) {
  switch (action.type) {
    case LOAD_PERSON_DETAILS:
      const orgId = action.orgId || 'personal';
      const org = action.org || {};
      const currentOrg = state.allByOrg[orgId];
      return {
        ...state,
        allByOrg: {
          ...updateAllPersonInstances(state.allByOrg, action.person, true), // update existing people
          [orgId]: {
            // make sure person is added to specified org or create the org if it doesn't exist
            ...org,
            ...currentOrg,
            people: {
              ...(currentOrg ? currentOrg.people : {}),
              [action.person.id]: action.person,
            },
          },
        },
      };
    case UPDATE_PERSON_ATTRIBUTES:
      return {
        ...state,
        allByOrg: updateAllPersonInstances(
          state.allByOrg,
          action.updatedPersonAttributes,
        ),
      };
    case DELETE_PERSON:
      return {
        ...state,
        allByOrg: deletePersonInOrg(
          state.allByOrg,
          action.personId,
          action.personOrgId,
        ),
      };
    case GET_ORGANIZATION_PEOPLE:
      return loadPeople(state, action);
    case REQUESTS.GET_MY_CHALLENGES.SUCCESS:
      return loadContactsFromSteps(state, action);
    case LOGOUT:
      return initialState;
    case PEOPLE_WITH_ORG_SECTIONS:
      return {
        ...state,
        allByOrg: action.orgs,
      };
    default:
      return state;
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function loadPeople(state: PeopleState, action: any) {
  const { response, orgId } = action;

  const org = state.allByOrg[orgId] || { id: orgId };
  const allPeople = org.people || {};
  response.forEach((person: Person) => {
    const existing = allPeople[person.id];
    allPeople[person.id] = existing ? { ...existing, ...person } : person;
  });

  return {
    ...state,
    allByOrg: {
      ...state.allByOrg,
      [orgId]: {
        ...org,
        people: allPeople,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function loadContactsFromSteps(state: PeopleState, action: any) {
  const { response } = action.results;

  const { allByOrg } = state;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  response.forEach((s: any) => {
    if (!s) {
      return;
    }
    const orgId = (s.organization && s.organization.id) || 'personal';
    const receiver = s.receiver;
    if (!receiver) {
      return;
    }

    if (!allByOrg[orgId]) {
      allByOrg[orgId] = { people: {} };
    }

    const existingContact = allByOrg[orgId].people[receiver.id] || {};

    allByOrg[orgId].people[receiver.id] = {
      ...existingContact,
      ...receiver,
    };
  });

  return {
    ...state,
    allByOrg,
  };
}

function updateAllPersonInstances(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  allByOrg: any,
  updatedPerson: Person,
  replace = false,
) {
  return mapObject(allByOrg, ([orgId, org]: [string, Organization]) => ({
    [orgId]: {
      ...org,
      people: {
        ...org.people,
        ...(org.people[updatedPerson.id]
          ? {
              ...(replace
                ? { [updatedPerson.id]: updatedPerson }
                : {
                    [updatedPerson.id]: {
                      ...org.people[updatedPerson.id],
                      ...updatedPerson,
                    },
                  }),
            }
          : {}),
      },
    },
  }));
}

function deletePersonInOrg(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  allByOrg: any,
  deletePersonId: string,
  personOrgId = 'personal',
) {
  return mapObject(allByOrg, ([orgId, org]: [string, Organization]) => ({
    [orgId]: {
      ...org,
      people: {
        ...(orgId === personOrgId
          ? filterObject(
              org.people,
              ([personId]: [string]) => personId !== deletePersonId,
            )
          : org.people),
      },
    },
  }));
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function mapObject(obj: any, fn: (element: any) => any) {
  return Object.assign({}, ...Object.entries(obj).map(fn));
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function filterObject(obj: any, fn: (element: any) => any) {
  return Object.assign(
    {},
    ...Object.entries(obj)
      .filter(fn)
      .map(([id, element]) => ({ [id]: element })),
  );
}
