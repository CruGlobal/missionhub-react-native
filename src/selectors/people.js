import { createSelector } from 'reselect';

import { removeHiddenOrgs } from './selectorUtils';

export const peopleByOrgSelector = createSelector(
  ({ people }) => people.allByOrg,
  ({ auth }) => auth.person,
  (orgs, authUser) =>
    sortOrgs(removeHiddenOrgs(Object.values(orgs), authUser), authUser)
      .map(org => ({
        ...org,
        people: Object.values(org.people)
          .filter(person => isAssignedToMeInOrganization(person, org, authUser))
          .sort((a, b) => {
            // Sort people in org by first name, then last name
            // Keep "ME" person in front
            if (a.id === authUser.id) {
              return -1;
            }
            if (b.id === authUser.id) {
              return 1;
            }
            return (
              a.first_name.localeCompare(b.first_name) ||
              a.last_name.localeCompare(b.last_name)
            );
          }),
      }))
      .filter(o => o.people && (o.id === 'personal' || o.people.length > 0)),
);

const isAssignedToMeInOrganization = (person, org, appUserPerson) => {
  const { reverse_contact_assignments } = person;

  if (person.id === appUserPerson.id) {
    return true;
  }

  return (
    reverse_contact_assignments &&
    reverse_contact_assignments.filter(a => {
      const { assigned_to, organization } = a;

      return (
        assigned_to &&
        assigned_to.id === appUserPerson.id &&
        (!organization || organization.id === org.id)
      );
    }).length > 0
  );
};

const sortOrgs = (orgs, authUser) => {
  const orgOrder = authUser.user.organization_order;

  return orgOrder
    ? sortWithPersonalInFront(
        orgs,
        (a, b) => orgOrder.indexOf(a.id) > orgOrder.indexOf(b.id),
      )
    : sortWithPersonalInFront(
        orgs,
        (a, b) => (a.name ? a.name.localeCompare(b.name) : 1),
      );
};

const sortWithPersonalInFront = (orgs, sortFn) =>
  orgs.sort((a, b) => {
    // Sort orgs by name
    // Keep Personal Ministry org in front
    if (a.id === 'personal') {
      return -1;
    }
    if (b.id === 'personal') {
      return 1;
    }
    return sortFn(a, b);
  });

export const personSelector = createSelector(
  ({ people }) => people.allByOrg,
  (_, { orgId }) => orgId,
  (_, { personId }) => personId,
  (orgs, orgId, personId) => {
    const org = orgs[orgId || 'personal'];
    return org && org.people[personId];
  },
);

export const contactAssignmentSelector = createSelector(
  (_, { person }) => person,
  (_, { orgId }) => orgId,
  ({ auth }) => auth.person.id,
  (person, orgId, authUserId) =>
    person.reverse_contact_assignments &&
    person.reverse_contact_assignments.find(
      assignment =>
        assignment.assigned_to &&
        assignment.assigned_to.id === authUserId &&
        (!assignment.organization || orgId === assignment.organization.id) &&
        (!orgId ||
          person.organizational_permissions.some(
            org_permission =>
              assignment.organization &&
              org_permission.organization_id === assignment.organization.id,
          )),
    ),
);

export const orgPermissionSelector = createSelector(
  (_, { person }) => person,
  (_, { organization }) => organization,
  (person, organization) =>
    organization &&
    (person.organizational_permissions || []).find(
      orgPermission => orgPermission.organization_id === organization.id,
    ),
);

export const contactsInOrgSelector = createSelector(
  ({ people }, { organization }) =>
    (people.allByOrg[organization.id] &&
      people.allByOrg[organization.id].people) ||
    {},
  (_, { organization }) => organization.contacts || [],
  (orgPeople, contactIds) => contactIds.map(i => orgPeople[i]),
);

export const contactsInSurveySelector = createSelector(
  ({ people }, { organization }) =>
    (people.allByOrg[organization.id] &&
      people.allByOrg[organization.id].people) ||
    {},
  (_, { organization, surveyId }) =>
    (organization.surveys &&
      organization.surveys[surveyId] &&
      organization.surveys[surveyId].contacts) ||
    [],
  (orgPeople, contactIds) => contactIds.map(i => orgPeople[i]),
);
