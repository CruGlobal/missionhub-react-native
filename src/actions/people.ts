import { PEOPLE_WITH_ORG_SECTIONS } from '../constants';
import { REQUESTS } from '../api/routes';
import { Person } from '../reducers/people';

import callApi from './api';

export function getMyPeople() {
  // @ts-ignore
  return async (dispatch, getState) => {
    const peopleQuery = {
      filters: {
        assigned_tos: 'me',
      },
      page: {
        limit: 1000,
      },
      include:
        'reverse_contact_assignments,reverse_contact_assignments.organization,organizational_permissions',
    };

    const loadedPeople: Person[] = (
      await dispatch(callApi(REQUESTS.GET_PEOPLE_LIST, peopleQuery))
    ).response;
    const authPerson = getState().auth.person;

    const people = [
      authPerson,
      ...loadedPeople.filter(person => {
        return person.reverse_contact_assignments.some(
          // @ts-ignore
          contactAssignment =>
            contactAssignment.assigned_to &&
            contactAssignment.assigned_to.id === authPerson.id,
        );
      }),
    ];

    return dispatch({ type: PEOPLE_WITH_ORG_SECTIONS, response: people });
  };
}

export function searchPeople(text = '', filters = {}) {
  // @ts-ignore
  return dispatch => {
    const query = {
      q: text,
      fields: {
        person: 'first_name,last_name',
        organization: 'name',
      },
      include:
        'organizational_permissions.organization,reverse_contact_assignments',
      filters: {},
    };
    // @ts-ignore
    if (filters.ministry) {
      // @ts-ignore
      query.organization_ids = filters.ministry.id;
    }
    // @ts-ignore
    if (filters.gender) {
      // @ts-ignore
      query.filters.gender = filters.gender.id;
    }
    // @ts-ignore
    if (filters.archived) {
      // @ts-ignore
      query.filters.archived = true;
    }
    // @ts-ignore
    if (filters.unassigned) {
      // @ts-ignore
      query.filters.unassigned = true;
    }
    // @ts-ignore
    if (filters.labels) {
      // @ts-ignore
      query.filters.label_ids = filters.labels.id;
    }
    // @ts-ignore
    if (filters.groups) {
      // @ts-ignore
      query.filters.group_ids = filters.groups.id;
    }
    // @ts-ignore
    if (filters.surveys) {
      // @ts-ignore
      query.filters.survey_ids = filters.surveys.id;
    }
    // @ts-ignore
    if (filters.permission_ids) {
      // @ts-ignore
      query.filters.permission_ids = filters.permission_ids;
    }

    return dispatch(callApi(REQUESTS.SEARCH, query));
  };
}
