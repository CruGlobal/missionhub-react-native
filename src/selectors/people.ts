import { createSelector } from 'reselect';

import { PeopleState, Person } from '../reducers/people';
import { AuthState } from '../reducers/auth';
import { Organization } from '../reducers/organizations';
import { CelebrateItem_subjectPerson_communityPermissions_nodes as CommunityPermission } from '../components/CelebrateItem/__generated__/CelebrateItem';

import { removeHiddenOrgs } from './selectorUtils';

interface Org {
  id: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  people: Person[];
  name: string;
}

export const peopleByOrgSelector = createSelector(
  ({ people }: { auth: AuthState; people: PeopleState }) => people.allByOrg,
  ({ auth }: { auth: AuthState; people: PeopleState }) => auth.person,
  (orgs, authUser) =>
    sortOrgs(removeHiddenOrgs(Object.values(orgs), authUser), authUser)
      .map((org: Org) => ({
        ...org,
        people: Object.values(org.people)
          .filter(person => isAssignedToMeInOrganization(person, org, authUser))
          .sort((a, b) => sortPeople(a, b, authUser)),
      }))
      .filter(
        (o: Org) => o.people && (o.id === 'personal' || o.people.length > 0),
      ),
);

export const allAssignedPeopleSelector = createSelector(
  ({ people }: { auth: AuthState; people: PeopleState }) => people.allByOrg,
  ({ auth }: { auth: AuthState; people: PeopleState }) => auth.person,
  (orgs, authUser) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const allPeople: any = {};
    removeHiddenOrgs(Object.values(orgs), authUser).forEach((org: Org) => {
      Object.values(org.people)
        .filter((person: Person) =>
          isAssignedToMeInOrganization(person, org, authUser),
        )
        .forEach((person: Person) => {
          allPeople[person.id] = {
            ...(allPeople[person.id] || {}),
            ...person,
          };
        });
    });
    return (Object.values(allPeople) || []).sort((a, b) =>
      sortPeople(a, b, authUser),
    );
  },
);

const isAssignedToMeInOrganization = (
  person: Person,
  org: Org,
  appUserPerson: Person,
) => {
  const { reverse_contact_assignments } = person;

  if (person.id === appUserPerson.id && org.id === 'personal') {
    return true;
  }

  return (
    reverse_contact_assignments &&
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    reverse_contact_assignments.filter((a: any) => {
      const { assigned_to, organization } = a;

      return (
        assigned_to &&
        assigned_to.id === appUserPerson.id &&
        (!organization || organization.id === org.id)
      );
    }).length > 0
  );
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const sortOrgs = (orgs: any, authUser: Person) => {
  const orgOrder = authUser.user.organization_order;

  return orgOrder
    ? sortWithPersonalInFront(
        orgs,
        (a: Org, b: Org) => orgOrder.indexOf(a.id) > orgOrder.indexOf(b.id),
      )
    : sortWithPersonalInFront(orgs, (a: Org, b: Org) =>
        a.name ? a.name.localeCompare(b.name) : 1,
      );
};

const sortWithPersonalInFront = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  orgs: any,
  sortFn: (a: Org, b: Org) => boolean | number,
) =>
  orgs.sort((a: Org, b: Org) => {
    // Sort orgs by name
    // Keep Personal Ministry org in front
    if (a.id === 'personal') {
      return -1;
    }
    if (b.id === 'personal') {
      return 1;
    }
    return sortFn(a, b);
  });

const sortPeople = (a: Person, b: Person, authUser: Person) => {
  // Sort people in org by first name, then last name
  // Keep "ME" person in front
  if (a.id === authUser.id) {
    return -1;
  }
  if (b.id === authUser.id) {
    return 1;
  }
  return (
    a.first_name.localeCompare(b.first_name) ||
    a.last_name.localeCompare(b.last_name)
  );
};

export const personSelector = createSelector(
  ({ people }: { people: PeopleState }) => people.allByOrg,
  (
    _: { people: PeopleState },
    { orgId }: { personId: string; orgId?: string },
  ) => orgId,
  (
    _: { people: PeopleState },
    { personId }: { personId: string; orgId?: string },
  ) => personId,
  (orgs, orgId, personId) => {
    const org = orgs[orgId || 'personal'];
    return org && org.people[personId];
  },
);

export const contactAssignmentSelector = createSelector(
  (_: { auth: AuthState }, { person }: { person: Person; orgId?: string }) =>
    person || {},
  (_: { auth: AuthState }, { orgId }: { person: Person; orgId?: string }) =>
    orgId,
  ({ auth }: { auth: AuthState }) => auth.person.id,
  (person, orgId, authUserId) => {
    const {
      reverse_contact_assignments = [],
      organizational_permissions = [],
    } = person;

    return reverse_contact_assignments.find(
      (assignment: {
        assigned_to?: { id: string };
        organization?: { id: string };
      }) =>
        assignment.assigned_to &&
        assignment.assigned_to.id === authUserId &&
        (!orgId || orgId === 'personal'
          ? !assignment.organization
          : assignment.organization &&
            orgId === assignment.organization.id &&
            organizational_permissions.some(
              (org_permission: { organization_id: string }) =>
                org_permission.organization_id ===
                (assignment.organization && assignment.organization.id),
            )),
    );
  },
);

export const orgPermissionSelector = createSelector(
  (_: {}, { person }: { person: Person; organization: Organization }) => person,
  (_: {}, { organization }: { person: Person; organization: Organization }) =>
    organization,
  (person, organization) =>
    organization && person.organizational_permissions
      ? (person.organizational_permissions || []).find(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          (orgPermission: any) =>
            orgPermission.organization_id === organization.id,
        )
      : (
          (person.communityPermissions && person.communityPermissions.nodes) ||
          []
        ).find(
          (orgPermission: CommunityPermission) =>
            orgPermission.community.id === organization.id,
        ),
);
