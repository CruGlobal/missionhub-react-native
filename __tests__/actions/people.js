import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { searchPeople, getMyPeople } from '../../src/actions/people';
import callApi, { REQUESTS } from '../../src/actions/api';

jest.mock('../../src/actions/api');

const mockStore = configureStore([thunk]);
let store;

const myId = 23;
const mockUser = {
  id: myId,
  name: 'Test User',
};

describe('getMyPeople', () => {
  const peopleQuery = {
    filters: {
      assigned_tos: 'me',
    },
    page: {
      limit: 1000,
    },
    include:
      'reverse_contact_assignments,reverse_contact_assignments.organization,organizational_permissions',
  };

  describe('as Casey', () => {
    const peopleList = [
      {
        id: 123,
        organizational_permissions: [],
        reverse_contact_assignments: [{ assigned_to: { id: myId } }],
      },
    ];

    it('should return one org with people', async () => {
      callApi.mockReturnValue({
        type: REQUESTS.GET_PEOPLE_LIST.SUCCESS,
        response: peopleList,
      });
      store = mockStore({
        auth: {
          isJean: false,
          person: mockUser,
        },
      });

      await store.dispatch(getMyPeople());

      expect(callApi).toHaveBeenCalledWith(
        REQUESTS.GET_PEOPLE_LIST,
        peopleQuery,
      );
      expect(store.getActions()[1]).toMatchSnapshot();
    });
  });

  describe('as Jean', () => {
    const organizationOne = {
      id: 101,
      name: 'organizationOne',
    };
    const organizationTwo = {
      id: 111,
      name: 'organizationTwo',
    };
    const personOne = {
      id: 7777,
      organizational_permissions: [{ organization: organizationOne }],
      reverse_contact_assignments: [
        { organization: { id: 103 }, assigned_to: { id: myId } },
        { organization: organizationOne, assigned_to: { id: myId } },
      ],
    };
    const personTwo = {
      id: 8888,
      organizational_permissions: [{ organization: { id: 102 } }],
      reverse_contact_assignments: [
        { organization: organizationOne, assigned_to: { id: myId } },
      ],
    };
    const personThree = {
      id: 9999,
      organizational_permissions: [],
      reverse_contact_assignments: [{ assigned_to: { id: myId } }],
    };
    const personFour = {
      id: 10000,
      organizational_permissions: [
        { organization: organizationTwo },
        { organization: organizationOne },
      ],
      reverse_contact_assignments: [
        { organization: organizationTwo, assigned_to: { id: myId } },
        { organization: organizationOne, assigned_to: { id: 24 } },
      ],
    };
    const personFive = {
      id: myId,
      organizational_permissions: [{ organization: organizationTwo }],
      reverse_contact_assignments: [
        { organization: organizationTwo, assigned_to: { id: myId } },
      ],
    };

    it('should return all orgs with assigned people', async () => {
      store = mockStore({
        auth: {
          isJean: true,
          person: mockUser,
        },
      });
      callApi.mockReturnValue({
        type: REQUESTS.GET_PEOPLE_LIST.SUCCESS,
        response: [personOne, personTwo, personThree, personFour, personFive],
      });

      await store.dispatch(getMyPeople());
      expect(callApi).toHaveBeenCalledWith(
        REQUESTS.GET_PEOPLE_LIST,
        peopleQuery,
      );
      expect(store.getActions()[1]).toMatchSnapshot();
    });
  });
});

describe('search', () => {
  const text = 'test';
  const inputFilters = {
    ministry: {
      id: '1',
    },
    gender: {
      id: '2',
    },
    archived: {
      id: '3',
    },
    unassigned: {
      id: '4',
    },
    labels: {
      id: '5',
    },
    groups: {
      id: '6',
    },
    surveys: {
      id: '7',
    },
    question: {
      id: '8',
      answer: {
        text: 'test',
      },
    },
  };

  const expectedQuery = {
    q: text,
    fields: {
      person: 'first_name,last_name',
      organization: 'name',
    },
    include: 'organizational_permissions.organization',
    filters: {},
  };
  const action = { type: 'ran search' };

  beforeEach(() => {
    store = mockStore();
    callApi.mockReturnValue(action);
  });

  const expectedFilterQuery = {
    ...expectedQuery,
    q: '',
    organization_ids: inputFilters.ministry.id,
    filters: {
      gender: inputFilters.gender.id,
      archived: true,
      unassigned: true,
      label_ids: inputFilters.labels.id,
      group_ids: inputFilters.groups.id,
      survey_ids: inputFilters.surveys.id,
      question_id: inputFilters.question.id,
      answer_value: inputFilters.question.answer.text,
    },
  };

  it('should search', () => {
    store.dispatch(searchPeople(text));

    expect(callApi).toHaveBeenCalledWith(REQUESTS.SEARCH, expectedQuery);
    expect(store.getActions()[0]).toBe(action);
  });

  it('should search with filters', () => {
    store.dispatch(searchPeople('', inputFilters));

    expect(callApi).toHaveBeenCalledWith(REQUESTS.SEARCH, expectedFilterQuery);
    expect(store.getActions()[0]).toBe(action);
  });
});
