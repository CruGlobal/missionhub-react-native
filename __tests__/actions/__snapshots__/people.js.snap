// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMyPeople as Casey should return one org with people 1`] = `
Array [
  Object {
    "response": Array [
      Object {
        "id": 123,
        "organizational_permissions": Array [],
        "reverse_contact_assignments": Array [
          Object {
            "assigned_to": Object {
              "id": 23,
            },
          },
        ],
      },
    ],
    "type": "GET_PEOPLE_LIST_SUCCESS",
  },
  Object {
    "orgs": Object {
      "personal": Object {
        "id": "personal",
        "people": Object {
          "123": Object {
            "id": 123,
            "organizational_permissions": Array [],
            "reverse_contact_assignments": Array [
              Object {
                "assigned_to": Object {
                  "id": 23,
                },
              },
            ],
          },
          "23": Object {
            "id": 23,
            "name": "Test User",
          },
        },
      },
    },
    "type": "app/PEOPLE_WITH_ORG_SECTIONS",
  },
]
`;

exports[`getMyPeople as Jean should return all orgs with assigned people 1`] = `
Array [
  Object {
    "response": Array [
      Object {
        "id": 7777,
        "organizational_permissions": Array [
          Object {
            "organization": Object {
              "id": 101,
            },
          },
        ],
        "reverse_contact_assignments": Array [
          Object {
            "assigned_to": Object {
              "id": 23,
            },
            "organization": Object {
              "id": 103,
            },
          },
          Object {
            "assigned_to": Object {
              "id": 23,
            },
            "organization": Object {
              "id": 101,
            },
          },
        ],
      },
      Object {
        "id": 8888,
        "organizational_permissions": Array [
          Object {
            "organization": Object {
              "id": 102,
            },
          },
        ],
        "reverse_contact_assignments": Array [
          Object {
            "assigned_to": Object {
              "id": 23,
            },
            "organization": Object {
              "id": 101,
            },
          },
        ],
      },
      Object {
        "id": 9999,
        "organizational_permissions": Array [],
        "reverse_contact_assignments": Array [
          Object {
            "assigned_to": Object {
              "id": 23,
            },
          },
        ],
      },
      Object {
        "id": 10000,
        "organizational_permissions": Array [
          Object {
            "organization": Object {
              "id": 111,
            },
          },
        ],
        "reverse_contact_assignments": Array [
          Object {
            "assigned_to": Object {
              "id": 23,
            },
            "organization": Object {
              "id": 111,
            },
          },
          Object {
            "assigned_to": Object {
              "id": 24,
            },
            "organization": Object {
              "id": 111,
            },
          },
        ],
      },
    ],
    "type": "GET_PEOPLE_LIST_SUCCESS",
  },
  Object {
    "orgs": Object {
      "101": Object {
        "id": 101,
        "people": Object {
          "7777": Object {
            "id": 7777,
            "organizational_permissions": Array [
              Object {
                "organization": Object {
                  "id": 101,
                },
              },
            ],
            "reverse_contact_assignments": Array [
              Object {
                "assigned_to": Object {
                  "id": 23,
                },
                "organization": Object {
                  "id": 103,
                },
              },
              Object {
                "assigned_to": Object {
                  "id": 23,
                },
                "organization": Object {
                  "id": 101,
                },
              },
            ],
          },
        },
      },
      "111": Object {
        "id": 111,
        "people": Object {
          "10000": Object {
            "id": 10000,
            "organizational_permissions": Array [
              Object {
                "organization": Object {
                  "id": 111,
                },
              },
            ],
            "reverse_contact_assignments": Array [
              Object {
                "assigned_to": Object {
                  "id": 23,
                },
                "organization": Object {
                  "id": 111,
                },
              },
              Object {
                "assigned_to": Object {
                  "id": 24,
                },
                "organization": Object {
                  "id": 111,
                },
              },
            ],
          },
        },
      },
      "personal": Object {
        "id": "personal",
        "people": Object {
          "23": Object {
            "id": 23,
            "name": "Test User",
          },
          "9999": Object {
            "id": 9999,
            "organizational_permissions": Array [],
            "reverse_contact_assignments": Array [
              Object {
                "assigned_to": Object {
                  "id": 23,
                },
              },
            ],
          },
        },
      },
    },
    "type": "app/PEOPLE_WITH_ORG_SECTIONS",
  },
]
`;
